cmd/sonarExecuteScan.go

```diff
@@ -0,0 +1,216 @@
+	}
+}
+
+func runSonar(options sonarExecuteScanOptions, client piperhttp.Downloader, runner execRunner) error {
+	if len(options.Host) > 0 {
+		sonar.Environment = append(sonar.Environment, "SONAR_HOST_URL="+options.Host)
+	}
+	if len(options.Token) > 0 {
+		sonar.Environment = append(sonar.Environment, "SONAR_AUTH_TOKEN="+options.Token)
+	}
+	if len(options.Organization) > 0 {
+		sonar.Options = append(sonar.Options, "sonar.organization="+options.Organization)
+	}
+	if len(options.ProjectVersion) > 0 {
+		sonar.Options = append(sonar.Options, "sonar.projectVersion="+options.ProjectVersion)
+	}
+	if err := handlePullRequest(options); err != nil {
+		return err
+	}
+	if err := loadSonarScanner(options.SonarScannerDownloadURL, client); err != nil {
+		return err
+	}
+	if err := loadCertificates(options.CustomTLSCertificateLinks, client, runner); err != nil {
+		return err
+	}
+
+	log.Entry().
+		WithField("command", sonar.Binary).
+		WithField("options", sonar.Options).
+		WithField("environment", sonar.Environment).
+		Debug("Executing sonar scan command")
+
+	sonar.Options = SliceUtils.Prefix(sonar.Options, "-D")
+	runner.SetEnv(sonar.Environment)
+	return runner.RunExecutable(sonar.Binary, sonar.Options...)
+}
+
+func handlePullRequest(options sonarExecuteScanOptions) error {
+	if len(options.ChangeID) > 0 {
+		if options.LegacyPRHandling {
+			// see https://docs.sonarqube.org/display/PLUG/GitHub+Plugin
+			sonar.Options = append(sonar.Options, "sonar.analysis.mode=preview")
+			sonar.Options = append(sonar.Options, "sonar.github.pullRequest="+options.ChangeID)
+			if len(options.GithubAPIURL) > 0 {
+				sonar.Options = append(sonar.Options, "sonar.github.endpoint="+options.GithubAPIURL)
+			}
+			if len(options.GithubToken) > 0 {
+				sonar.Options = append(sonar.Options, "sonar.github.oauth="+options.GithubToken)
+			}
+			if len(options.Owner) > 0 && len(options.Repository) > 0 {
+				sonar.Options = append(sonar.Options, "sonar.github.repository="+options.Owner+"/"+options.Repository)
+			}
+			if options.DisableInlineComments {
+				sonar.Options = append(sonar.Options, "sonar.github.disableInlineComments="+strconv.FormatBool(options.DisableInlineComments))
+			}
+		} else {
+			// see https://sonarcloud.io/documentation/analysis/pull-request/
+			provider := strings.ToLower(options.PullRequestProvider)
+			if provider == "github" {
+				sonar.Options = append(sonar.Options, "sonar.pullrequest.github.repository="+options.Owner+"/"+options.Repository)
+			} else {
+				return errors.New("Pull-Request provider '" + provider + "' is not supported!")
+			}
+			sonar.Options = append(sonar.Options, "sonar.pullrequest.key="+options.ChangeID)
+			sonar.Options = append(sonar.Options, "sonar.pullrequest.base="+options.ChangeTarget)
+			sonar.Options = append(sonar.Options, "sonar.pullrequest.branch="+options.ChangeBranch)
+			sonar.Options = append(sonar.Options, "sonar.pullrequest.provider="+provider)
```

Comment:
Just a suggestion: You could add a method `appendOption(option string)` to `sonarSettings` and safe on a lot of code-bloat:
```
sonar.appendOption("sonar.pullrequest.key="+options.ChangeID)
```
versus
```
sonar.Options = append(sonar.Options, "sonar.pullrequest.key="+options.ChangeID)
```

Reply:
good idea